apps/web/src/server/applications/create/applicant/applications-create-applicant.controller.js

```diff
@@ -135,27 +142,93 @@ export async function updateApplicationsCreateApplicantEmail(
 /**
  * View the form step for the applicant address
  *
- * @type {import('@pins/express').RenderHandler<{}, {}>}
+ * @type {import('@pins/express').RenderHandler<ApplicationsCreateApplicantAddressProps,
+  {}, {}, {postcode: string}, {}>}
  */
-export async function viewApplicationsCreateApplicantAddress(req, response) {
-	response.render('applications/create/applicant/_address');
+export async function viewApplicationsCreateApplicantAddress({ query }, response) {
+	const { postcode } = query;
+	const formStage = postcode ? 'manualAddress' : 'searchPostcode';
+
+	response.render('applications/create/applicant/_address', { formStage, postcode });
 }
 
 /**
  * Update the applicant address
  *
- * @type {import('@pins/express').RenderHandler<{}, {}>}
+ * @type {import('@pins/express').RenderHandler<ApplicationsCreateApplicantAddressProps,
+ * {}, ApplicationsCreateApplicantAddressBody, {}, {}>}
  */
-export async function updateApplicationsCreateApplicantAddress({ path, session }, response) {
-	const { applicationId } = response.locals;
+export async function updateApplicationsCreateApplicantAddress(
+	{ path, session, errors: validationErrors, body },
+	response
+) {
+	const { postcode, apiReference, currentFormStage } = body;
+	const { applicationId, applicantId } = response.locals;
+
+	const {
+		/** @type {ValidationErrors} */ errors: apiErrors,
+		/** @type {ApplicationsAddress[]} */ addressList: matchingAddressList,
+		/** @type {ApplicationCreateApplicantAddressStage} */ formStage = currentFormStage
+	} = await (async () => {
+		switch (currentFormStage) {
+			case 'searchPostcode': {
+				const { errors, addressList } = await findAddressListByPostcode(postcode, {
+					maxResults: 50,
+					minMatch: 0.9
+				});
+				const newFormStage = errors ? 'searchPostcode' : 'selectAddress';
+
+				return { errors, addressList, formStage: newFormStage };
+			}
+			case 'selectAddress': {
+				const { errors: serviceErrors, addressList } = await findAddressListByPostcode(postcode, {
```

Comment:
do you need to call `findAddressListByPostcode` also when you show the results?

Reply:
yes, we need to show them the addresses, to compare the selected apiReference with the list and finally to actually send the data to the db.
We could save them in the session in the previous step but I'm trying to use the session as little as possible and then I was thinking that we'll need to probably make a request anyway when we resume it